src/metabase/util/stats.clj

```diff
@@ -0,0 +1,332 @@
+(defn- database-dims 
+  "characterize a database record"
+  [database]
+  {:total 1
+   :analysed (if (database :is_full_sync) 1 0)}
+  )
+
+(defn- get-database-metrics 
+  "Get metrics based on databases"
+  []
+  (let [databases (db/select 'Database)]
+    {:databases (apply add-summaries (map database-dims databases))}))
+
+
+(defn- get-table-metrics 
+  "Get metrics based on tables"
+  []
+  (let [tables (db/select 'Table)]
+    {:tables (count tables)
+     :num_per_database (medium-histogram tables :db_id)
+     :num_per_schema (medium-histogram tables :schema)
+     }))
+
+
+(defn- get-field-metrics 
+  "Get metrics based on fields"
+  []
+  (let [fields (db/select 'Field)]
+    {:fields (count fields)
+     :num_per_table (medium-histogram fields :table_id)}))
+
+
+
+(defn- get-segment-metrics 
+  "Get metrics based on segments"
+  []
+  (let [segments (db/select 'Segment)]
+    {:segments (count segments)}))
+
+
+(defn- get-metric-metrics 
+  "Get metrics based on metrics"
+  []
+  (let [metrics (db/select 'Metric)]
+    {:metrics (count metrics)}))
+
+(defn- bin-latencies 
+  "Bin latencies, which are in milliseconds"
+  [query-executions]
+  (let [latency-vals (map #(/ %1 1000) (map :running_time query-executions))]
+    (frequencies (map bin-large-number latency-vals))))
+
+;; Execution Metrics
+(defn- get-execution-metrics 
+  "Get metrics based on executions. 
+  This should be done in a single pass, as there might 
+  be a LOT of query executions in a normal instance
+  TODO: characterize by ad hoc vs cards"
+  []
+  (let [executions (db/select 'QueryExecution)]
+    {:executions (count executions)
+     :by_status (frequencies (map :status executions))
+     :num_per_user (large-histogram executions :executor_id)
+     :num_per_card (large-histogram executions :table_id)
+     :num_by_latency (bin-latencies executions)}))
+
+(defn- get-map-metrics 
+  "Get metrics based on custom geojson maps
+  TODO figure out how to get at these"
+  []
+  (let [maps (db/select 'Segment)]
+    {:maps (count maps)}))
+
+
+(defn get-anonymous-usage-stats 
+  "generate a map of the usage stats for this instance"
+  []
+  (merge (get-settings)
+           {:uuid anonymous-id :timestamp (new java.util.Date)}
+            {:stats {
+             :user (get-user-metrics)
+             :question (get-question-metrics)
+             :dashboard (get-dashboard-metrics)
+             :database (get-database-metrics)
+             :table (get-table-metrics)
+             :field (get-field-metrics)
+             :pulse (get-pulse-metrics)
+             :segment (get-segment-metrics)
+             :metric (get-metric-metrics)
+             :group (get-group-metrics)
+             :label (get-label-metrics)
+             :execution (get-execution-metrics)}}))
+
+(defn- send-stats! 
+  "send stats to Metabase tracking server"
+  [stats]
+   (try 
+     (print (client/post metabase-usage-url {:form-params stats :content-type :json :throw-entire-message? true}))
```

Comment:
Maps on a single line should have key-value pairs separated by commas, e.g.

```clojure
{:form-params stats, :content-type :json, :throw-entire-message? true}
```

See https://github.com/bbatsov/clojure-style-guide#opt-commas-in-map-literals